runtime/src/iree/hal/drivers/cuda/cuda_device.c

```diff
@@ -234,6 +237,80 @@ static iree_status_t iree_hal_cuda_device_query_i64(
       (int)category.size, category.data, (int)key.size, key.data);
 }
 
+// Returns true if |id| is all zeros indicating an empty ID.
+static bool iree_hal_cuda_nccl_id_is_empty(const iree_hal_cuda_nccl_id_t* id) {
+  for (iree_host_size_t i = 0; i < IREE_ARRAYSIZE(id->data); ++i) {
+    if (id->data[i] != 0) return false;
+  }
+  return true;
+}
+
+static iree_status_t iree_hal_cuda_device_create_channel(
+    iree_hal_device_t* base_device, iree_hal_queue_affinity_t queue_affinity,
+    iree_hal_channel_params_t params, iree_hal_channel_t** out_channel) {
+  iree_hal_cuda_device_t* device = iree_hal_cuda_device_cast(base_device);
+
+  // TODO(#9580): check if nccl symbols are available - if not then we fail
+  // here and have the error propagated up to users. If we wanted to delay load
+  // NCCL we'd want to take a lock here, load it, and merge the symbols into the
+  // dynamic symbol table.
+  if (true) {
+    return iree_make_status(
+        IREE_STATUS_UNIMPLEMENTED,
+        "NCCL unavailable and collective operations cannot be performed");
+  }
+
+  // Try to use the ID specified in the parameters and fall back to the default.
+  iree_hal_cuda_nccl_id_t id;
+  if (iree_const_byte_span_is_empty(params.id)) {
+    // User wants the default.
+    id = device->params.nccl_default_id;
+  } else if (params.id.data_length != IREE_ARRAYSIZE(id.data)) {
+    // User provided something but it's not what we expect.
+    return iree_make_status(
+        IREE_STATUS_INVALID_ARGUMENT,
+        "NCCL ID must be %d bytes matching the ncclUniqueId struct",
+        (int)IREE_ARRAYSIZE(id.data));
+  } else {
+    // User provided the ID - we treat it as opaque here and let NCCL validate.
+    memcpy(id.data, params.id.data, IREE_ARRAYSIZE(id.data));
+  }
+  if (iree_hal_cuda_nccl_id_is_empty(&id)) {
+    // TODO: maybe this is ok? a localhost alias or something?
```

Comment:
I think it should be okay until we add a support for an automatic server bootstrapping, i.e., starting a server, getting the IP address, and broadcasting the IP address to the other process.

Reply:
yeah, for the broadcast stuff we will likely still want that to happen in the hosting application layer (PJRT/etc) and have them pass that in as the device default ID - or let the user create the channels themselves and pass those in via setters (`func.func @set_channel(%channel: !hal.channel)` etc) - I'm not sure we want to be in the business of MPI coordination at this layer but could have helpers for testing/etc in iree/tooling/